File: android/app/src/main/java/com/clipsync/android/logging/FileLogger.kt
package com.clipsync.android.logging

import android.content.Context
import java.io.File
import java.io.FileWriter
import java.io.PrintWriter
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/**
 * Simple file logger. Writes to app-internal storage.
 * Never logs clipboard content — only type, size, and truncated hashes.
 */
object FileLogger {
    private lateinit var logDir: File
    private val dateFormat = SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS", Locale.US)
    private val fileDateFormat = SimpleDateFormat("yyyyMMdd", Locale.US)
    private val lock = Any()

    fun init(context: Context) {
        logDir = File(context.filesDir, "logs")
        logDir.mkdirs()
        cleanOldLogs()
    }

    fun getLogDir(): File = logDir

    fun info(message: String) = log("INFO", message)
    fun warn(message: String) = log("WARN", message)
    fun error(message: String) = log("ERROR", message)
    fun error(message: String, throwable: Throwable) = log("ERROR", "$message: ${throwable.stackTraceToString()}")

    fun log(level: String, message: String) {
        synchronized(lock) {
            try {
                val today = fileDateFormat.format(Date())
                val file = File(logDir, "clipsync-$today.log")
                FileWriter(file, true).use { writer ->
                    val timestamp = dateFormat.format(Date())
                    writer.appendLine("[$timestamp] [$level] $message")
                }
            } catch (e: Exception) {
                // Best effort — don't crash over logging
            }
        }
    }

    /**
     * Returns all log content as a string (most recent file first).
     */
    fun getAllLogs(): String {
        return try {
            val files = logDir.listFiles { f -> f.name.startsWith("clipsync-") && f.name.endsWith(".log") }
                ?.sortedByDescending { it.name }
                ?: return "No logs found."
            if (files.isEmpty()) return "No logs found."
            buildString {
                for ((index, file) in files.withIndex()) {
                    if (index > 0) appendLine()
                    appendLine("--- ${file.name} ---")
                    appendLine()
                    append(file.readText())
                }
            }
        } catch (e: Exception) {
            "Error reading logs: ${e.message}"
        }
    }

    /**
     * Returns the most recent log entries (last N lines from today's file).
     */
    fun getRecentLogs(maxLines: Int = 200): String {
        return try {
            val today = fileDateFormat.format(Date())
            val file = File(logDir, "clipsync-$today.log")
            if (!file.exists()) return "No logs for today."
            val lines = file.readLines()
            val start = maxOf(0, lines.size - maxLines)
            lines.subList(start, lines.size).joinToString("\n")
        } catch (e: Exception) {
            "Error reading logs: ${e.message}"
        }
    }

    private fun cleanOldLogs() {
        try {
            val cutoff = System.currentTimeMillis() - 7 * 24 * 60 * 60 * 1000L
            logDir.listFiles { f -> f.name.startsWith("clipsync-") && f.lastModified() < cutoff }
                ?.forEach { it.delete() }
        } catch (_: Exception) {}
    }
}

File: windows/ClipSync.Windows/Logging/FileLogger.cs
using System.IO;
using System.Globalization;

namespace ClipSync.Windows.Logging;

/// <summary>
/// Simple file logger with daily rotation. Writes to %APPDATA%\ClipSync\logs\.
/// Never logs clipboard content — only type, size, and truncated hashes.
/// </summary>
public sealed class FileLogger : IDisposable
{
    private static readonly Lazy<FileLogger> _instance = new(() => new FileLogger());
    public static FileLogger Instance => _instance.Value;

    private readonly string _logDir;
    private StreamWriter? _writer;
    private string _currentDate = "";
    private readonly object _lock = new();

    private FileLogger()
    {
        _logDir = Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData),
            "ClipSync", "logs");
        Directory.CreateDirectory(_logDir);
    }

    public string LogDirectory => _logDir;

    public void Log(string level, string message)
    {
        lock (_lock)
        {
            EnsureWriter();
            var timestamp = DateTime.Now.ToString("HH:mm:ss.fff", CultureInfo.InvariantCulture);
            _writer!.WriteLine($"[{timestamp}] [{level}] {message}");
            _writer.Flush();
        }
    }

    public void Info(string message) => Log("INFO", message);
    public void Warn(string message) => Log("WARN", message);
    public void Error(string message) => Log("ERROR", message);
    public void Error(string message, Exception ex) => Log("ERROR", $"{message}: {ex}");

    private void EnsureWriter()
    {
        var today = DateTime.Now.ToString("yyyyMMdd", CultureInfo.InvariantCulture);
        if (today == _currentDate && _writer != null) return;

        _writer?.Dispose();
        _currentDate = today;
        var path = Path.Combine(_logDir, $"clipsync-{today}.log");
        _writer = new StreamWriter(path, append: true) { AutoFlush = true };

        // Clean up logs older than 7 days
        CleanOldLogs();
    }

    private void CleanOldLogs()
    {
        try
        {
            var cutoff = DateTime.Now.AddDays(-7);
            foreach (var file in Directory.GetFiles(_logDir, "clipsync-*.log"))
            {
                if (File.GetLastWriteTime(file) < cutoff)
                {
                    File.Delete(file);
                }
            }
        }
        catch
        {
            // Best effort
        }
    }

    public void Dispose()
    {
        lock (_lock)
        {
            _writer?.Dispose();
            _writer = null;
        }
    }
}
